v4: Simplify button states

mdo opened the pull request:
This PR drastically simplifies our button states to address #19189, #21235, and #21415. My hope is we can simplify the selectors, reduce the number of selectors overall, and make it easier to customize our buttons. Here's a rundown of what's changed and why.

- **Moved active state after the disabled state.** This way, disabled *active* buttons automatically inherit the proper changes to visual styles and simply modifies the opacity. Previously we had to reset the `background-color` and `border-color` across all states.

- **Unchains the focus state from active state.** There's no need to add extra selectors to the original focus state within `&, &:active, &.active { ... }`—we just need to set the focus styles.

- **Removes all hover, focus, and disabled styles from the active states.** Now that active is after disabled in Sass, we can easily inherit the proper styles. Similarly, hover and focus will inherit automatically as well. This does remove the ability to see a hover on active, but I'm okay with that as it's more akin to native behaviors.

See the new CSS in action at https://output.jsbin.com/homiyos.

/cc @cvrebert @patrickhlauke

mdo committed:
fixes #21415 by moving disabled .btn-link color to the default state

mdo committed:
Redo button states

Trying to simplify our output here by revamping these selectors. We overcomplicated things by setting hover styles for nearly every state (disabled and active included), and we set them in the wrong order.

This commit does the following:

- Reorders states so disabled comes before active, thereby removing the need to set disabled-active styles.

- Removes all focus and hover styles from disabled states as those will naturally inherit from the default button state.

- Renamed `.open` to `.show` to fix dropdown toggle highlighting.

- Tweaked some indendation in the Sass.

mdo committed:
linting

patrickhlauke commented:
not had a chance to test this out live, but just looking at the description cold:

> Unchains the focus state from active state

> Removes all hover, focus, and disabled styles from the active states

is it still possible to see a difference between a focused active and focused inactive buttons? if not, then i'd strongly suggest to reconsider this

mdo committed:
redo outlines

- removes original outline removal (hah)
- replaces it with an explicit `outline: 0` on `.btn`
- instead of replicating `:hover` for `:focus`, uses custom and themed `box-shadow` for an "outline"
- not mapped to the `$enable-shadows` variable because accessibility

mdo committed:
remove cursor: pointer

shoutout https://medium.com/simple-human/buttons-shouldnt-have-a-hand-cursor-b11e99ca374b#.riy12ck0m

mdo commented:
Instead of a different `background-color` and `border-color`, focused buttons have a new "outline" via `box-shadow`. I wish I could use `outline`, but most browsers don't round it to match `border-radius` and it causes the `background-color` to bleed out passed the rounded corners. 😢 

So, instead we have `box-shadow: 0 0 0 2px rgba($border, .5);`, which looks pretty good I think. Originally I was planning on having a pixel of negative space around the button (like an `outline-offset`, which no one supports), but that's difficult without setting another opaque shadow (e.g., `#fff`). As such, we have this custom shadow.

The benefit of these new selectors, @patrickhlauke, is that the `:focus` styles are automatically inherited into focused active buttons.

<img width="619" alt="screen shot 2016-12-26 at 10 10 29 pm" src="https://cloud.githubusercontent.com/assets/98681/21492315/722b6982-cbb8-11e6-9b21-6f67debde66c.png">

patrickhlauke commented:
@mdo on that screenshot, the "Focus state" and "Active focus state" are indistinguishable. assuming it's a mistake in the screenshot, right?

subjectively, i'd say the focus outline (shadow) could do with being slightly more pronounced to make it more obvious.

on a positive note, removing the hover state at least reduces the number of iterations, so that's definitely a good thing.

mdo committed:
remove active styling from focus

mdo commented:
@patrickhlauke Gah, fixed the screenshot and demo. I had left in the active bg and what not on the regular `:focus` state for the button variant mixin. Fixed now.

I'll also say that the focus outline _is_ more apparent when tabbing through, as opposed to everything being highlighted here in the demo. That said, I'll play with making the colors more vibrant here.

mdo committed:
create a variable for the focus box-shadow, use it in the button base class

also be sure to apply it for the active state when shadows are enabled

mdo committed:
properly pass focus shadow should you enable shadows

don't implement for the outline style because outline buttons get no shadow by default

md5 commented:
I just noticed the change introduced by caef376 as part of this PR, namely to remove the `cursor: pointer` from buttons.

While I understand the logic behind this change and I support the removal of that style, it introduced an inconsistency between `button.btn` and `a.btn` in that the former now has an arrow cursor and the latter has a hand cursor. It seems to me that the intent of applying `.btn` to an `<a>` tag is to make it look, act, and feel like a button, so I would think the right fix would be to set `cursor: default` on `.btn` instead of removing the `cursor` property entirely.

I can open a separate issue or provide a PR if that makes sense, but I wanted to start the discussion in the context where the original change was made.

md5 commented:
I've opened https://github.com/twbs/bootstrap/pull/22693 to make the change I'm suggesting.
